src/main/kotlin/org/wycliffeassociates/otter/common/domain/content/Recordable.kt

```diff
@@ -0,0 +1,54 @@
+package org.wycliffeassociates.otter.common.domain.content
+
+import org.wycliffeassociates.otter.common.data.model.ContentType
+import org.wycliffeassociates.otter.common.data.workbook.*
+
+class Recordable private constructor(
+    val textItem: TextItem? = null,
+    val audio: AssociatedAudio,
+    val start: Int? = null,
+    val end: Int? = null,
+    val sort: Int? = null,
+    val contentType: ContentType? = null
+) {
+    companion object {
+        fun build(bookElement: BookElement, component: Resource.Component): Recordable {
+            return when (bookElement) {
+                is Chapter -> Recordable(
+                    textItem = component.textItem,
+                    audio = component.audio,
+                    sort = component.sort,
+                    contentType = component.contentType
+                )
+                is Chunk -> Recordable(
+                    textItem = component.textItem,
+                    audio = component.audio,
+                    start = bookElement.start,
+                    end = bookElement.end,
+                    sort = component.sort,
+                    contentType = component.contentType
+                )
+                else -> throw IllegalStateException("Unsupported book element type found in Recordable.build")
+            }
+        }
+
+        fun build(chunk: Chunk) = Recordable(
+            textItem = chunk.text,
+            audio = chunk.audio,
+            start = chunk.start,
+            end = chunk.end
+        )
+    }
+
+    override fun equals(other: Any?): Boolean {
```

Comment:
This is used so that we know whether the user selected a new Recordable, so we don't trigger onChange events by setting a Recordable to the same thing
that was selected before

Reply:
if you're overriding equals you probably need to override hashcode

Reply:
Actually, overriding hashcode is a lot easier than I expected since I can just use `Objects.hash`

Reply:
This looks off to me. I've never written one myself but when I've seen articles and they all follow a similar looking pattern. I'd prefer to defer to the expertise in the Effective Java book which I believe I saw on Craig's desk. Could you take a look at it?

Reply:
Objects.hash should work just fine. It uses the hashCode() implementation of each parameter that is passed to Objects.hash. For a little more info...
https://stackoverflow.com/questions/45832458/use-objects-hash-or-own-hashcode-implementation

Reply:
I just realized that the only classes that implement Recordable are data classes, for which Kotlin already provides an implementation for `equals()` and `hashCode()`. 

While we can't really guarantee that data classes are the only classes that will ever implement Recordable, I personally don't think it's worth creating equals() and hashcode() implementations in Recordable. If we do this, classes that implement Recordable still have to override equals() and hashcode() and call Recordable's default implementation of equals() or hashcode() in the function body.

If we want to guarantee that objects that implement Recordable are data classes, we would have to turn Recordable into a data class itself rather than an interface, and have Chunk and Resource.Component extend it, but this seems like overkill to me.